li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaFederatedProducerImpl.java

```diff
@@ -355,7 +361,68 @@ public void close() {
 
   @Override
   public void close(long timeout, TimeUnit timeUnit) {
-    throw new UnsupportedOperationException("Not implemented yet");
+    if (_producers.isEmpty()) {
+      LOG.warn("No producers to close for cluster group {}", _clusterGroup);
+      return;
+    }
+
+    LOG.info("Closing LiKafkaProducer for cluster group {} in {} {}...", _clusterGroup, timeout, timeUnit);
+
+    long startTimeMs = System.currentTimeMillis();
+    long budgetMs = timeUnit.toMillis(timeout);
+    long deadlineTimeMs = startTimeMs + budgetMs;
+
+    List<Thread> taskThreads = new ArrayList<>();
+    for (Map.Entry<ClusterDescriptor, PerClusterProducerContext<K, V>> entry : _producers.entrySet()) {
+      Thread taskThread = new Thread(new CloseTask(entry.getValue(), deadlineTimeMs));
+      taskThreads.add(taskThread);
+      taskThread.start();
+    }
+
+    for (Thread taskThread : taskThreads) {
+      try {
+        taskThread.join();
```

Comment:
Would a countdown latch or cyclic barrier work better here? the for-loop imposes some kind of implicit order in which the threads terminate

Reply:
I think functionality-wise, using join should be fine (each thread is supposed to be independent and I don't think implicit ordering doesn't really matter). But I switched to countdown latch for better readability/simplicity.